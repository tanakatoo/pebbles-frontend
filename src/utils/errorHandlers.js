import { rest } from 'msw'

let loginResponse = {
    token: 'abcdefg',
    user: {
        id: 2,
        username: 'hello',
        role: 'regular',
        premium_join_date: '',
        premium_end_date: '',
        free_trial_start_date: '',
    }
}



export const errorHandlers = [

    //login
    rest.post('http://localhost:3001/auth/login', (req, res, ctx) => {
        const failLogin = {
            "error": {
                "message": "INVALID_CREDENTIALS",
                "status": 401
            }
        }
        console.log(res(ctx.json([failLogin])))
        return res(ctx.json([failLogin]));
    }),

    rest.post('http://localhost:3001/email', (req, res, ctx) => {
        console.log('MOCK error /email');
        const noData = {
            "error": {
                "message": "no data",
                "status": 400
            }
        }
        return res(ctx.json(noData));
    }),

]