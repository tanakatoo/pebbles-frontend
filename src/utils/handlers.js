import { rest } from 'msw'

let loginResponse = {
    token: 'abcdefg',
    user: {
        id: 2,
        username: 'hello',
        role: 'regular',
        premium_join_date: '',
        premium_end_date: '',
        free_trial_start_date: '',
    }
}

const contactsResponse =
    [
        {
            "user_id": 1,
            "username": "ktoo",
            "avatar": "2.jpg"
        },
        {
            "user_id": 15,
            "username": "june",
            "avatar": "4.jpg"
        }
    ]

const blockedResponse = [
    {
        "blocked_user_id": 3,
        "username": "blockMe",
        "avatar": "9.jpg"
    }
]
const mainSearchResults = [
    {
        "username": "hello2",
        "avatar": "9.jpg",
        "name": null,
        "about": null,
        "premium_acct_id": null,
        "end_date": null
    },
    {
        "username": "helloo",
        "avatar": "14.jpg",
        "name": null,
        "about": null,
        "premium_acct_id": null,
        "end_date": null
    }
]

const changePassword = "completed"

const newMessage = [
    {
        "toavatar": "9.jpg",
        "to": "hello",
        "fromavatar": "",
        "from": "hello"
    }
]

const continuingMsg = [
    {
        "toavatar": "9.jpg",
        "fromavatar": "2.jpg",
        "from": "ktoo",
        "to": "hello",
        "from_user_id": 1,
        "to_user_id": 2,
        "msg": "the most recent one reply to second one",
        "sent_at": "2023-06-09 15:30:00",
        "read": true
    },
    {
        "toavatar": "2.jpg",
        "fromavatar": "9.jpg",
        "from": "hello",
        "to": "ktoo",
        "from_user_id": 2,
        "to_user_id": 1,
        "msg": "2nd most recent one reply",
        "sent_at": "2023-06-08 15:30:00",
        "read": true
    },
    {
        "toavatar": "9.jpg",
        "fromavatar": "2.jpg",
        "from": "ktoo",
        "to": "hello",
        "from_user_id": 1,
        "to_user_id": 2,
        "msg": "first message",
        "sent_at": "2023-06-07 15:30:00",
        "read": true
    }
]

const latestMessages = [
    {
        "toavatar": "9.jpg",
        "fromavatar": "4.jpg",
        "from": "june",
        "to": "hello",
        "from_user_id": 15,
        "to_user_id": 2,
        "msg": "testing\n",
        "sent_at": "2023-07-15 20:57:44.804634",
        "read": false
    },
    {
        "toavatar": "9.jpg",
        "fromavatar": "2.jpg",
        "from": "ktoo",
        "to": "hello",
        "from_user_id": 1,
        "to_user_id": 2,
        "msg": "the most recent one reply to second one",
        "sent_at": "2023-06-09 15:30:00",
        "read": true
    }
]

const returnUser = [
    {
        "id": 2,
        "username": "hello",
        "name": null,
        "avatar": "9.jpg",
        "country_en": null,
        "country_ja": null,
        "city_en": null,
        "city_ja": null,
        "state_en": null,
        "state_ja": null,
        "gender": "female",
        "about": null,
        "free_trial_start_date": null,
        "study_buddy_bio": "just chatting",
        "study_buddy_purpose": "i don't know why",
        "native_language": "Japanese",
        "learning_language": "English",
        "language_level": "Beginner",
        "time_zone": "NewYork",
        "age_range": "26-35",
        "study_buddy_active": true,
        "study_buddy_activate_date": "2023-07-13T04:00:00.000Z"
    }
]

export const handlers = [

    //login
    rest.post('http://localhost:3001/auth/login', (req, res, ctx) => {

        console.log(res(ctx.json(loginResponse)))
        return res(ctx.json(loginResponse));
    }),

    rest.post('http://localhost:3001/auth/logout', (req, res, ctx) => {

        console.log('MOCK /logout')
        window.localStorage.setItem('token', null)
        return res(ctx.json());
    }),


    // Called via <ExploreCommunity type="studyBuddy" />
    rest.get('http://localhost:3001/study-buddies/1', (req, res, ctx) => {
        console.log('MOCK /study-buddies/1');

        return res(ctx.json([]));
    }),


    rest.post('http://localhost:3001/auth/register', (req, res, ctx) => {
        console.log('MOCK /auth/register');

        return res(ctx.json(loginResponse));
    }),


    rest.post('http://localhost:3001/email', (req, res, ctx) => {
        console.log('MOCK /email');

        return res(ctx.json('sents'));
    }),


    rest.get('http://localhost:3001/users/contacts', (req, res, ctx) => {
        console.log('MOCK /users/contacts');

        return res(ctx.json(contactsResponse));
    }),

    rest.get('http://localhost:3001/users/blocked', (req, res, ctx) => {
        console.log('MOCK /users/blocked');

        return res(ctx.json(blockedResponse));
    }),

    rest.get('http://localhost:3001/users/search', (req, res, ctx) => {
        console.log('MOCK /users/search');

        return res(ctx.json(mainSearchResults));
    }),

    rest.post('http://localhost:3001/auth/change-password', (req, res, ctx) => {
        console.log('MOCK /auth/change-password');

        return res(ctx.json(changePassword));
    }),

    //for continuing conversation
    rest.get('http://localhost:3001/messages/:username', (req, res, ctx) => {
        console.log('MOCK /messages/:username');
        console.log(req.params)
        return res(ctx.json(continuingMsg));
    }),

    //get latest message from all users
    rest.get('http://localhost:3001/messages', (req, res, ctx) => {
        console.log('MOCK /messages');
        return res(ctx.json(latestMessages));
    }),

    //get latest message from all users
    rest.get('http://localhost:3001/users/saved-users', (req, res, ctx) => {
        console.log('MOCK /users/saved-users');
        return res(ctx.json(returnUser));
    }),
]