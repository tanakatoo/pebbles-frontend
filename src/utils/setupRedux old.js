
import { configureStore } from '@reduxjs/toolkit';
import rootReducer from "../reducers/rootReducer"
import { persistStore, persistReducer, FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER } from 'redux-persist';

import storage from 'redux-persist/lib/storage'
import autoMergeLevel2 from 'redux-persist/lib/stateReconciler/autoMergeLevel2'

export const persistConfig = {
    key: 'root', // Key prefix for the persisted state
    storage, // Storage mechanism (e.g., local storage)
    stateReconciler: autoMergeLevel2, // Specify which parts of the state to persist
    whitelist: ['profile', 'langFont']
};

export const persistedReducer = persistReducer(persistConfig, rootReducer);
export const store = configureStore({
    reducer: persistedReducer,
    middleware: (getDefaultMiddleware) =>
        getDefaultMiddleware({
            serializableCheck: {
                ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER]
            }
        })
})
export const persistedStore = persistStore(store)