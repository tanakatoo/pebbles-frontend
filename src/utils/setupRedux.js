import React from "react";
import { render } from "@testing-library/react";
import { configureStore } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
// As a basic setup, import your same slice reducers
import { persistStore, persistReducer, FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER } from 'redux-persist';
import { PersistGate } from 'redux-persist/integration/react'
import { ListSlice } from "../reducers/rootReducer";
import rootReducer from "../reducers/rootReducer"
import storage from 'redux-persist/lib/storage'
import storeSetup from "./storeSetup"
import autoMergeLevel2 from 'redux-persist/lib/stateReconciler/autoMergeLevel2'

// const persistConfig = {
//     key: 'root', // Key prefix for the persisted state
//     storage, // Storage mechanism (e.g., local storage)
//     stateReconciler: autoMergeLevel2, // Specify which parts of the state to persist
//     whitelist: ['profile', 'langFont']
// };
// const persistedReducer = persistReducer(persistConfig, rootReducer);

// const persistedStore = persistStore(store)
// const root = ReactDOM.createRoot(document.getElementById('root'));


export function renderWithProviders(
    ui,
    {
        preloadedState = {},
        // Automatically create a store instance if no store was passed in
        store = storeSetup(preloadedState),
        ...renderOptions
    } = {}
) {
    function Wrapper({ children }) {

        return <Provider store={store}>{children}</Provider>;
    }

    // Return an object with the store and all of RTL's query functions
    return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}