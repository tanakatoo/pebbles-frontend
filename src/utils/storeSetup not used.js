import { combineReducers, configureStore } from '@reduxjs/toolkit'

import rootReducer from '../reducers/rootReducer'
// import userReducer from '../features/users/userSlice'

// Create the root reducer separately so we can extract the RootState type
// const rootReducer = combineReducers({
//     user: userReducer
// })

export const setupStore = preloadedState => {
    return configureStore({
        reducer: rootReducer,
        preloadedState
    })
}