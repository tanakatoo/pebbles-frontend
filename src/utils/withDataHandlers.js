import { rest } from 'msw'

let studybuddyResponse = [
    {
        "id": 2,
        "username": "hello",
        "name": null,
        "avatar": "9.jpg",
        "country_en": null,
        "country_ja": null,
        "city_en": null,
        "city_ja": null,
        "state_en": null,
        "state_ja": null,
        "gender": "female",
        "about": null,
        "free_trial_start_date": null,
        "study_buddy_bio": "just chatting",
        "study_buddy_purpose": "i don't know why",
        "native_language": "Japanese",
        "learning_language": "English",
        "language_level": "Beginner",
        "time_zone": "NewYork",
        "age_range": "26-35",
        "study_buddy_active": true,
        "study_buddy_activate_date": "2023-07-13T04:00:00.000Z",
        "study_buddy_types": [
            "StudyBuddy",
            "Volunteer"
        ]
    }
]

console.log('in data handlers', studybuddyResponse)

export const withDataHandlers = [

    //get some study buddies
    rest.get('http://localhost:3001/study-buddies/1', (req, res, ctx) => {
        console.log('MOCK with data /study-buddies/1')
        console.log(studybuddyResponse)
        return res(ctx.json(studybuddyResponse));
    }),

]