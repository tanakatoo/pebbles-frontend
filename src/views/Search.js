import React, { useEffect, useRef, useState } from 'react'
import SearchBar from '../components/common/SearchBar'
import useFormData from '../hooks/useFormData'
import { useLocation } from 'react-router-dom'
import { Tab } from '@headlessui/react'
import usePageText from '../hooks/usePageText'
import Card from "../components/studyBuddy/Card"
import { v4 as uuid } from "uuid"
import useNavigateToProfile from '../hooks/useNavigateToProfile'
import ServerError from '../components/form/ServerError'
import Spinner from '../components/common/Spinner'
import UserApi from '../api/user'
import UsersResultsRight from '../components/form/search/UsersResultsRight'
import UsersResultsTopRight from '../components/form/search/UsersResultsTopRight'
import { Formik, Form } from 'formik'

function Search() {
    const location = useLocation()
    // const [data, setData, handleChange, resetData] = useFormData({ word: location.state.word })
    const formRef = useRef()
    const [pageText, lang] = usePageText('searchBar')
    const [selectedTabIndex, setSelectedTabIndex] = useState(0)
    const tabs = [pageText.USERS, pageText.INFO_CENTER, pageText.LANGUAGE_TOWN, pageText.MARKETPLACE]
    const goToProFile = useNavigateToProfile()
    const [doneGettingData, setDoneGettingData] = useState(true)
    const [results, setResults] = useState(null)
    const [errors, setErrors] = useState([])

    const initialValues = { word: '' }
    let formValues = {}
    if (location.state && location.state.from === '/') {
        formValues = { word: location.state.word }
    }

    useEffect(() => {
        window.scrollTo(0, 0);
        if (location.state && location.state.from === '/') {
            formRef.current.handleSubmit();
        }
    }, [])


    const submitSearch = async (values, { setSubmitting }) => {

        if (values.word === '') {

            setDoneGettingData(true)
            setResults(null)
            setErrors([])
        } else {
            setDoneGettingData(false)
            setErrors([])
            try {
                const res = await UserApi.findUsers({ word: values.word })

                setResults(res)
            } catch (e) {
                if (e instanceof TypeError) {
                    //means server is down
                    console.error('TypeError saving edit profile data', e)
                    setErrors(["UNKNOWN"])
                } else {
                    console.error('Error saving edit profile data', e)
                    setErrors(e)
                }
            } finally {
                setSubmitting(false)
                setDoneGettingData(true)
            }


        }
    }
    const handleKeyDown = (e, formik) => {

        if (e.key === 'Enter') {

            submitSearch(formik.values, formik)
        }
    }

    return (
        <div className='Search'>
            <div className="bg-primary-dark py-6 px-4">
                <Formik
                    initialValues={formValues || initialValues}
                    onSubmit={submitSearch}
                    innerRef={formRef}
                    enableReinitialize

                >{formik => {

                    return (
                        <Form data-testid="searchForm" onKeyDown={(e) => handleKeyDown(e, formik)}>
                            <SearchBar
                                placeholder={pageText.PLACEHOLDER}
                                name='word'
                                data={formik.values.word}
                                onSubmit={submitSearch}
                            />
                        </Form>
                    )
                }}
                </Formik>
            </div>
            {errors.length > 0 && <ServerError msg={errors} />}
            {!results && errors.length === 0 && doneGettingData === false ?
                <div className=' my-24'>
                    <Spinner />
                </div>
                :
                results === null && errors.length === 0 && doneGettingData === true ?


                    <div className='my-24 max-w-prose mx-auto w-full'>
                        <p className='text-center mb-8'>
                            {pageText.MSG_1}
                        </p>
                        <p className='text-center mb-8'> {pageText.MSG_2}</p>
                        <p className='text-center'> {pageText.MSG_3}</p>
                    </div>

                    : results.length === 0 && errors.length === 0 && doneGettingData === true ?
                        <div className='my-24 max-w-prose mx-auto w-full'>
                            <p className='text-center mb-8'>
                                {pageText.NO_INFO}
                            </p>
                            <p className='text-center mb-8'>{pageText.NO_INFO2}</p>

                        </div>
                        :
                        <div className='w-full mt-6'>
                            <Tab.Group selectedIndex={selectedTabIndex} onChange={setSelectedTabIndex}>
                                <Tab.List className="flex justify-between flex-col md:flex-row">
                                    <Tab className="w-full pb-1 md:pb-3 border-b-2 
                                        ui-not-selected:border-b-gray-stroke
                                        text-gray-text
                                        ui-selected:text-primary-dark
                                        ui-selected:font-bold
                                        ui-selected:outline-none">
                                        {tabs[0]}
                                    </Tab>

                                    <Tab className="w-full py-2 md:pb-3 border-b-2 
                                        ui-not-selected:border-b-gray-stroke
                                        text-gray-text
                                        ui-selected:font-bold
                                        ui-selected:text-primary-dark
                                        ui-selected:outline-none">
                                        {tabs[1]}
                                    </Tab>
                                    <Tab className="w-full py-2 md:pb-3 border-b-2 
                                        ui-not-selected:border-b-gray-stroke
                                        text-gray-text
                                        ui-selected:font-bold
                                        ui-selected:text-primary-dark
                                        ui-selected:outline-none">
                                        {tabs[2]}
                                    </Tab>
                                    <Tab className="w-full py-2 md:pb-3 border-b-2 
                                        ui-not-selected:border-b-gray-stroke
                                        text-gray-text
                                        ui-selected:font-bold
                                        ui-selected:text-primary-dark
                                        ui-selected:outline-none">
                                        {tabs[3]}
                                    </Tab>

                                </Tab.List>
                                <div className='my-6'>
                                    <Tab.Panels>
                                        <Tab.Panel>
                                            <div className=' mx-auto'>
                                                <div data-testid="results" className='mt-4 flex  flex-wrap justify-center mb-12 gap-4 px-2'>
                                                    {results.map(d =>
                                                        <Card data={d}
                                                            goToProfileOnClick={goToProFile}
                                                            key={uuid()}
                                                            buddy={false}
                                                            topRightStatic={true}
                                                            topRight={<UsersResultsTopRight data={d} />}
                                                            underUsername={<UsersResultsRight data={d} />}
                                                        />)}

                                                </div>
                                            </div>


                                        </Tab.Panel>

                                        <Tab.Panel>
                                            <div className='my-24 max-w-prose mx-auto w-full'>
                                                <p className='text-center mb-8'>
                                                    {pageText.INFO_CENTER1}
                                                </p>
                                                <p className='text-center mb-8'>{pageText.INFO_CENTER2}</p>
                                                <p className='text-center'>{pageText.CHECK_BACK}</p>
                                            </div>

                                        </Tab.Panel>
                                        <Tab.Panel>

                                            <div className='my-24 max-w-prose mx-auto w-full'>
                                                <p className='text-center mb-8'>
                                                    {pageText.TOWN1}
                                                </p>
                                                <p className='text-center mb-8 max-w-prose'>{pageText.TOWN2}</p>
                                                <p className='text-center'>{pageText.CHECK_BACK}</p>
                                            </div>
                                        </Tab.Panel>
                                        <Tab.Panel>
                                            <div className='my-24 max-w-prose mx-auto w-full'>
                                                <p className='text-center mb-8 '>
                                                    {pageText.MARKET1}
                                                </p>
                                                <p className='text-center mb-8 '>{pageText.MARKET2}</p>
                                                <p className='text-center'>{pageText.CHECK_BACK}</p>
                                            </div>
                                        </Tab.Panel>
                                    </Tab.Panels>
                                </div>
                            </Tab.Group>
                        </div>}
        </div>
    )
}

export default Search