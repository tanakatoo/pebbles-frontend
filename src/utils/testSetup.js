// import { persistedStore, store } from './setupRedux';
import { render } from '@testing-library/react'
import { Provider } from 'react-redux';
import { PersistGate } from 'redux-persist/integration/react'
import { setupTestStore } from './setupRedux';


const renderWithProviders = (ui, options = {}) => {
    const [store, persistedStore] = setupTestStore(options.preloadedState);

    const AllProviders = ({ children }) => {
        return (
            <Provider store={store}>
                <PersistGate loading={null} persistor={persistedStore}>
                    {children}
                </PersistGate>
            </Provider>
        )
    };

    return { store, ...render(ui, { wrapper: AllProviders, ...options }) };
}


// "re-export everything"
export * from '@testing-library/react';

export { renderWithProviders };