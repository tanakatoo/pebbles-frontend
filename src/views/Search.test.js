import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { renderWithProviders } from '../utils/testSetup';
import Search from './Search';
import { fireEvent, waitFor } from '../utils/testSetup';
import { act } from 'react-dom/test-utils';


// Mock window.scrollTo()
window.scrollTo = jest.fn();
afterAll(() => {
    jest.clearAllMocks();
});


describe('Search', () => {
    // test('renders without crashing', () => {
    //     renderWithProviders(
    //         <MemoryRouter>
    //             <Search />
    //         </MemoryRouter>
    //     );
    // });

    // test('matches snapshot', async () => {
    //     const { asFragment } = renderWithProviders(
    //         <MemoryRouter>
    //             <Search />
    //         </MemoryRouter>
    //     );

    //     expect(asFragment()).toMatchSnapshot();
    // });
    test('search results display', async () => {
        const { findByText, getByTestId, getByRole, getByText } = await renderWithProviders(
            <MemoryRouter>
                <Search />
            </MemoryRouter>
        );

        act(async () => {
            let elem = getByRole('textbox');
            fireEvent.change(elem, { target: { value: 'hello' } });
            fireEvent.keyDown(elem, { key: 'Enter', code: 'Enter', charCode: 13 })

            // Wait for pageText to be populated
            await waitFor(async () => {
                expect(getByTestId('results')).not.toBeEmptyDOMElement();

                await findByText('helloo')

            })
        })
    })

    // fireEvent.click(button)
    // await screen.findByText('Clicked twice')
    // // Submit form
    // await act(async () => {
    //     await fireEvent.keyDown(elem, { key: 'Enter', code: 'Enter', charCode: 13 })
    // })
    // // fireEvent.click(getByTestId(/loginFormLogin/));


    // });

    // expect(getByText(/helloo/)).toBeInTheDocument();

});


