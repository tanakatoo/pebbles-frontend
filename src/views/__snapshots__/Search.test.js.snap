// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="bg-primary-dark py-6 px-4"
    >
      <form
        action="#"
      >
        <div
          class="w-full flex  null gap-3"
        >
          <input
            class="rounded-ml grow placeholder-gray focus:ring-none"
            name="word"
            placeholder="Search"
            type="text"
            value=""
          />
        </div>
      </form>
    </div>
    <div
      class="my-24 max-w-prose mx-auto w-full"
    >
      <p
        class="text-center mb-8"
      >
        Use the search bar above to search for anything on the site.
      </p>
      <p
        class="text-center mb-8"
      >
         Currently the search only includes users, but in the future, items from the marketplace, resources and groups from Japanese Town and Regional Information Center will be displayed.
      </p>
      <p
        class="text-center"
      >
         Check back from time to time!
      </p>
    </div>
  </div>
</DocumentFragment>
`;
